### PR TITLE
Support arcs in sketches

### PR DESCRIPTION
**Pull request by antonok-edm**
closes https://github.com/hannobraun/Fornjot/issues/100

2023-01-05-104443_414x317_scrot
_screenshot of the `test` model as modified from this branch_

There are a few problems with model validation in some cases. I found that in particular, adding arcs to the 5-pointed star will cause a lot of errors. But from general testing on simple shapes, I've found that the arcs are quite reliable.

I'm happy to reorganize things here as necessary - I got it "working" but I didn't spend too long thinking about how/where to define things like `ArcCircleData`. I also might not be using the most idiomatic patterns in some places.

### CODE REVIEW

**Comment by antonok-edm** [comment]
I also didn't touch sketched `Circle`s here but it'd be pretty simple afterwards to represent them internally as a polychain with a single arc edge.

**Review by hannobraun — CHANGES_REQUESTED** [review]
Thank you, @antonok-edm, this is absolutely awesome!

I've left some comments with things that I noticed. Only one of those needs to be addressed (the potential bug in the bounding bugs calculation). Everything else is 100% optional and can be addressed in a follow-up pull request, or not at all.


> There are a few problems with model validation in some cases. I found that in particular, adding arcs to the 5-pointed star will cause a lot of errors. But from general testing on simple shapes, I've found that the arcs are quite reliable.

That's perfectly fine. My only concern are regressions to existing functionality. I'm perfectly happy to merge not-quite-reliable new features :smile: 

My main concern is the user experience, as it's quite easy to cause the app to crash by changing the angle value in the test model. Eventually, I'd like to be able to mark features as experimental, requiring users to opt into them explicitly (https://github.com/hannobraun/Fornjot/issues/431).

But until we have that ability, I'd rather merge new features and gain the benefits from them immediately. Keeping pull requests open long-term until they are "perfect" is not a good solution, in my opinion.


> I'm happy to reorganize things here as necessary - I got it "working" but I didn't spend too long thinking about how/where to define things like `ArcCircleData`.

I've left some comments with my thoughts, but as I said above, all of that is optional. More improvements are of course very welcome, but those can happen in follow-up PRs. If it weren't for the bounding box issue, I'd merge this immediately.


> I also might not be using the most idiomatic patterns in some places.

I was actually quite impressed by that builder code you added! I think a month from now, I won't be able to tell which parts of that builder I wrote :smile: 

As a general note though, what's considered idiomatic in the context of Fornjot is in flux, as I constantly discover better patterns. Nothing is set in stone, so in case you have any feedback, I'd love to hear it!


> I also didn't touch sketched `Circle`s here but it'd be pretty simple afterwards to represent them internally as a polychain with a single arc edge.

Also fine! These kinds of cleanups don't need to hold up this pull request.

---

Okay, so the next step to move this forward would be to address the bounding box calculation, which I think should be rather simple.

If you want to address any of my other comments at the same time (or have ideas of your own), feel free to do that. But we don't need to keep this PR open longer than necessary. Happy to merge this now, and make further improvements whenever is convenient.

File: crates/fj-operations/src/sketch.rs
Comment (antonok-edm) [reply crates/fj-operations/src/sketch.rs:101]: I struggled with this one for a bit actually. It can't go in the same match statement because both ends of each `HalfEdge` need to be part of the polychain before these methods will work. `update_as_line_segment` updates the direction, and `update_as_arc` calculates the center point, radius, and angles; all of these are wrong unless both endpoints (i.e. the one from _this_ edge and also the _next_ edge) are already defined.

There's probably a better way to structure this. Maybe just storing the segment kind inside of the `HalfEdge` as another partial type, and then calculating all of the other metrics later? But I think that's definitely a followup.

File: crates/fj-operations/src/sketch.rs
Comment (antonok-edm) [reply crates/fj-operations/src/sketch.rs:155]: hmm, I don't _think_ I've changed anything for the `SketchSegmentRoute::Direct` case?

Previously, it just collected all of the endpoints directly:

```rust
fj::Chain::PolyChain(poly_chain) => Aabb::<3>::from_points(
    poly_chain
        .to_segments()
        .into_iter()
        .map(|fj::SketchSegment::LineTo { point }| point)
        .map(Point::from)
        .map(Point::to_xyz),
),
```

In this PR, there's a bit of indirection but it's still collecting and processing them all the same way:

```rust
// (some lines irrelevant to `SketchSegmentRoute::Direct` removed)
fj::Chain::PolyChain(poly_chain) => {
    let segments = poly_chain.to_segments();
    let mut points = vec![];
    segments.iter().for_each(|segment| {
        points.push(Point::from(segment.endpoint));
    });

    Aabb::<3>::from_points(points.into_iter().map(Point::to_xyz))
}
```

File: crates/fj-kernel/src/builder/edge.rs
Comment (antonok-edm) [reply crates/fj-kernel/src/builder/edge.rs]: Yes, good idea - pushed an update for that. I also updated it to be consistent with the fact that it's accepting radians, not degrees.

File: crates/fj-kernel/src/geometry/path.rs
Comment (antonok-edm) [reply crates/fj-kernel/src/geometry/path.rs]: easy enough to fix in this PR too, updated

File: crates/fj-math/src/circle.rs
Comment (antonok-edm) [reply crates/fj-math/src/circle.rs]: Yeah, looking at it again, that makes sense. I initially imagined it to be a small addition to `Circle` but at this point it basically deserves its own module.

Took the chance to add some tests for it too, and discovered that angles with a magnitude greater than π radians were broken, so that's fixed now too :rocket:

File: crates/fj-operations/src/sketch.rs
Comment (hannobraun) [reply crates/fj-operations/src/sketch.rs:101]: Ah, got it. Looks like I was a bit naive in my initial judgement :smile: 


> There's probably a better way to structure this. Maybe just storing the segment kind inside of the `HalfEdge` as another partial type, and then calculating all of the other metrics later? But I think that's definitely a followup.

Let's just leave it as-is for now. The structure of the various object and partial object types, as well as their relationships to each other are somewhat in flux anyway. Maybe a more elegant solution will present itself, but until then, this will do nicely.

File: crates/fj-operations/src/sketch.rs
Comment (hannobraun) [reply crates/fj-operations/src/sketch.rs:173]: Oh yeah, I was looking at this wrong. I didn't see that `points.push` outside of the `match` and thought points were only pushed in the `Arc` case. Sorry!

And indeed, I can reproduce the missing corner on current `main`. Unfortunately not reliably, but it does seem to be an existing bug. I have to remember to open an issue for that.

Again, sorry! I misread the code, and then jumped to conclusions when I came across that bug in my testing.

File: crates/fj-kernel/src/geometry/path.rs
Comment (hannobraun) [reply crates/fj-kernel/src/geometry/path.rs]: Thanks!

File: crates/fj-math/src/circle.rs
Comment (hannobraun) [reply crates/fj-math/src/circle.rs]: Thank you, and nice catch finding that bug!

File: crates/fj-operations/src/sketch.rs
Comment (hannobraun) [reply crates/fj-operations/src/sketch.rs]: Ha, looks like the shorter variable name helped rustfmt sort itself out :smile:

**Review by hannobraun — APPROVED** [review]
Thank you for the follow-up, @antonok-edm!

As I said in my reply to your comment, that bounding box bug is actually pre-existing and I was misreading your code. Sorry for that!

This is good to go now. Merging!

Thanks again. It's great to have this new CAD modeling capability!

**Pull request merged**
